### ISSUE TITLE
Validatefunc is ignored when referencing a variable for a custom provider.

### ISSUE THREAD

**Issue by thomasmeeus** [hashicorp/terraform#28104]
Hi,
I'm writing a custom terraform provider and use `ValidateFunc` for validating input. This doesn't seem to work for referenced values?  The value is never validated and Terraform continues with possible incorrect values. When not using a reference to a variable, but having it inline validation works as intended.

### Terraform Version

```
terraform -version
Terraform v0.14.8
```

### Terraform Configuration Files

```terraform

variable "api_url"{
  default = "https://my-api-endpoint"
}

provider "my_custom_provider" {
  username = var.api_username
  password = var.api_password
  url      = var.api_url  // this value is never validated
}
```

This works as it should:
```terraform
provider "my_custom_provider" {
  username = var.api_username
  password = var.api_password
  url      = "https://my-api-endpoint"  // this value validated
}
```

### Expected Behavior
ValidateFunc should be called when using both direct input & referenced variables

### Actual Behavior
Referenced variables aren't validated by `ValidateFunc`

### References
- https://github.com/hashicorp/terraform/issues/21746

**Comment by apparentlymart** [comment]
Hi @thomasmeeus,

I think understanding this behavior requires some context that isn't generally obvious from the UI. Since you didn't include a trace log in your report I'm making some assumptions here based on my mental model of Terraform's behavior, but I think this should be accurate enough to start thinking about the behavior you are seeing.

The first thing that's important to consider here is that Terraform has a validate phase which aims to check whether the configuration itself is correct, regardless of what dynamic values might be set for variables at a later time. In that phase, root module input variables don't have values at all yet (technically, they are marked as "unknown") and so Terraform does only type checking of them: it'll catch if you try to assign a list variable where a string value is expected, for example, but it doesn't yet know the concrete value of `var.api_url` because that's an argument provided to the _plan_ operation. Terraform calls the provider API function which the SDK translates into `ValidateFunc` initially as part of this validate phase, at which point the `url` argument in your first example will have an unknown value of type string and so the SDK will skip calling `ValidateFunc` at that stage.

`terraform validate` runs the validation phase in isolation, while `terraform plan` runs the validation phase prior to creating the plan, because we need to know that the configuration is valid itself before we can consider the effect of it with particular input variables and provider results.

The intended design is that Terraform Core would call the provider's validate function again during the plan and apply operations, with each of those subsequent calls adding further information compared to what was known in the previous phase. A missing call to do that was, as far as I can tell, what fixed #21225 in #21555: it added the missing extra call to `provider.ValidateResourceTypeConfig` during the planning phase, which allows the SDK (on the provider's behalf) to re-validate the configuration in a `resource` block to include new information learned during the planning phase.

With all of this said, my first suspect for understanding what you saw here was to make sure Terraform has a corresponding call to re-validate the provider configuration itself (as opposed to individual resource configurations) during the planning phase. Unfortunately (or fortunately, depending on how you look at it), there does appear to be a call to re-validate the provider configuration just prior to configuring the provider, and so I don't think this problem can be explained in the same way as #21225 was:

https://github.com/hashicorp/terraform/blob/6697245c18dad6848e7647598dda1ab04c2680ca/terraform/node_provider.go#L134-L177

There doesn't seem to be any path to `provider.ConfigureProvider` (which the SDK translates into `provider.Configure` in your code) that doesn't first call `provider.ValidateProviderConfig` (which the SDK mostly handles itself, but does call your `ValidateFunc` implementations as part of that work.)

With that said then, I think it'll take some more information to explain what you're seeing here. One way in which the behavior you saw here can be _expected_ is if `var.api_url` is being populated from a resource attribute that won't be known until the apply step, because that value would be unknown even during planning. In order to understand if that applies to your situation, or if there's something else special going on that I'm not considering, it would help if you could share some additional information:

* The full configuration you used for testing, exactly as you ran it.
* The command or sequence of commands you ran while testing, up to and including the one where you observed the behavior you described here.
* The full debug log for the final step in your sequence of commands, which you can get by setting the environment variable `TF_LOG=trace` before running that Terraform CLI command. Trace logs are typically too long for GitHub comments, so it'll be easier to share it as a link to a [gist](https://gist.github.com/). 

With the above we should at least be able to understand the full sequence of operations Terraform is following in the situation you're describing here. We won't be able to reproduce it exactly because we don't have access to your custom provider, but hopefully the trace log will give enough information to construct an equivalent case with a provider we _do_ have access to, so we can observe the behavior and debug it better.

Ultimately it's the responsibility of [the SDK](https://github.com/hashicorp/terraform-plugin-sdk) to call your `ValidateFunc`, rather than Terraform Core itself calling it, but to understand why the SDK is behaving how it is we'll need to first understand how Terraform Core is behaving, to make sure it is indeed calling `ConfigureProvider` as the above code suggests and, if so, what values the provider arguments have at that point in the process.

Thanks!

**Comment by thomasmeeus** [comment]
Woah, I wasn't expecting such an extensive response! Thanks for that :-)

I'm basically writing a Terraform provider to communicate with our company specific configuration database. 
It's the first time writing a provider. It works fine, but it's still in an early stage.
I tried adding `ValidateFunc` in order to enhance the experience for the end-users. 
This issue isn't a deal breaker, more like a nice to have. Most of all I'm wondering what's going wrong inside Terraform.

Here are the details you requested. I had to ommit some company specific stuff, but I don't think that will make a difference.

ValidateFunc:

```
ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
    apiURL := val.(string)
    _, err := url.ParseRequestURI(apiURL)
    if err != nil {
        errs = append(errs, fmt.Errorf("The supplied value '%s' for input '%s' could not be parsed, error: %s. We expect the following: '[scheme:][//host]' ", apiURL, key, err.Error()))
    }
    return warns, errs
},
```

Example code with no validation. Maybe because there's no resource object defined:

```
variable "usd_api_username" {}
variable "usd_api_password" {}

variable "usd_api_url"{
  default = "example.com"
}

provider "usd" {
  username = var.usd_api_username
  password = var.usd_api_password
  url      = "example.com"
}

```

No validation. We reference a variable here:

```
variable "usd_api_username" {}
variable "usd_api_password" {}

variable "usd_api_url"{
  default = "example.com"
}

provider "usd" {
  username = var.usd_api_username
  password = var.usd_api_password
  url      = var.usd_api_url
}

resource "usd_ci" "my_second_customer_server" {
  organization     = "Customer"
  configuration    = "Customer_configuration"
  hostname         = "Example_hostname"
  resolvergroup    = "My_team"
  servicewindow    = "businesshours"
  oncall           = null
  cpu_count        = 2
  memory_count     = 2
  operating_system = "Redhat CoreOS (RHCOS)"
  role             = "openshift-master"
}
```

Working validation via direct input to the provider:

```
variable "usd_api_username" {}
variable "usd_api_password" {}

variable "usd_api_url"{
  default = "example.com" // this variable isn't used in this example
}

provider "usd" {
  username = var.usd_api_username
  password = var.usd_api_password
  url      = "example.com" 
}

resource "usd_ci" "my_second_customer_server" {
  organization     = "Customer"
  configuration    = "Customer_configuration"
  hostname         = "Example_hostname"
  resolvergroup    = "My_team"
  servicewindow    = "businesshours"
  oncall           = null
  cpu_count        = 2
  memory_count     = 2
  operating_system = "Redhat CoreOS (RHCOS)"
  role             = "openshift-master"
}
```

Commands I run to reproduce this:
```
terraform init
terraform plan # this returns either the validation error from my custom provider or an 'ugly' error from the underlaying Go httpClient library.
terraform apply # this obviously returns the same error as terraform plan. 
````

Here's the trace output:
* terraform-output-with-validation.txt: https://gist.github.com/thomasmeeus/9af50d3c69ee18dd782fd4cb39e28341
* terraform-output-no-validation.txt: https://gist.github.com/thomasmeeus/5e945f71a9fc00b1392715926a1c8164

Let me know if you need anything else.
Thanks!

**Comment by apparentlymart** [comment]
Thanks for sharing those details, @thomasmeeus! As an aid to whoever looks at this next, I wanted to extract what seem to be the most relevant parts of the `terraform-output-no-validation.txt` log related to the interactions between Terraform Core and the provider plugin:

```
2021/03/17 10:54:07 [TRACE] terraform.NewContext: starting
2021/03/17 10:54:07 [TRACE] terraform.NewContext: loading provider schemas
2021/03/17 10:54:07 [TRACE] LoadSchemas: retrieving schema for provider type "terraform.my_company.com/my_company/usd"
2021-03-17T10:54:07.396+0100 [DEBUG] plugin: using plugin: version=5
2021/03/17 10:54:07 [TRACE] GRPCProvider: GetSchema
2021/03/17 10:54:07 [TRACE] GRPCProvider: Close
2021/03/17 10:54:07 [TRACE] terraform.NewContext: complete
2021/03/17 10:54:07 [DEBUG] Starting graph walk: walkValidate
2021/03/17 10:54:07 [TRACE] vertex "provider[\"terraform.my_company.com/my_company/usd\"]": starting visit (*terraform.NodeApplyableProvider)
2021/03/17 10:54:07 [TRACE] buildProviderConfig for provider["terraform.my_company.com/my_company/usd"]: using explicit config only
2021/03/17 10:54:07 [TRACE] GRPCProvider: GetSchema
2021/03/17 10:54:07 [TRACE] GRPCProvider: PrepareProviderConfig
2021/03/17 10:54:07 [TRACE] vertex "provider[\"terraform.my_company.com/my_company/usd\"]": visit complete
2021/03/17 10:54:07 [TRACE] vertex "provider[\"terraform.my_company.com/my_company/usd\"] (close)": starting visit (*terraform.graphNodeCloseProvider)
2021/03/17 10:54:07 [TRACE] GRPCProvider: Close
2021/03/17 10:54:07 [TRACE] vertex "provider[\"terraform.my_company.com/my_company/usd\"] (close)": visit complete
2021/03/17 10:54:07 [DEBUG] Starting graph walk: walkPlan
2021/03/17 10:54:07 [TRACE] vertex "provider[\"terraform.my_company.com/my_company/usd\"]": starting visit (*terraform.NodeApplyableProvider)
2021/03/17 10:54:07 [TRACE] buildProviderConfig for provider["terraform.my_company.com/my_company/usd"]: using explicit config only
2021/03/17 10:54:07 [TRACE] GRPCProvider: GetSchema
2021/03/17 10:54:07 [TRACE] GRPCProvider: Configure
2021/03/17 10:54:07 [TRACE] vertex "provider[\"terraform.my_company.com/my_company/usd\"]": visit complete
2021/03/17 10:54:07 [TRACE] vertex "usd_ci.my_second_customer_server": dynamic subgraph encountered errors
2021/03/17 10:54:07 [TRACE] dag/walk: upstream of "provider[\"terraform.my_company.com/my_company/usd\"] (close)" errored, so skipping
```

From this subset, the thing that sticks out to me the most is that I see a line `GRPCProvider: PrepareProviderConfig` in the validate phase but _not_ in the plan phase, which doesn't match my expectations based on my intuition for how Terraform ought to be have here and also on my read of the code snippet I shared in my previous comment. The previous comment shows a call to `provider.ValidateProviderConfig` instead of to `provider.PrepareProviderConfig` because I took it from the latest main branch rather than from v0.14.8 and there have been some changes to the internal provider API in the meantime, so just for comparison's sake I'll find the equivalent codepath in v0.14.8 too:

https://github.com/hashicorp/terraform/blob/ee7197beaed55037819fdefc3b29404c810e1603/terraform/node_provider.go#L86-L123

Interestingly, this older version _doesn't_ include a call to `provider.PrepareProviderConfig`, and so this _might_ be a situation where the bug you saw was already fixed in the meantime, and so the behavior might well be better in a newer release.

After a little digging, I found #26816 which seems to be describing the same misbehavior you reported here. It looks like this is recent enough that so far it's only been included in Terraform v0.15.0-beta1, and not in any stable release, but if you are just testing a new provider under development rather than using Terraform in production perhaps you could give this a try with [a v0.15.0-beta1 package](https://releases.hashicorp.com/terraform/0.15.0-beta1/) and see if it behaves as you (and I) expected? (We're also planning to release beta2 shortly, so if that's there by the time you see this then feel free to try that one instead; it should also contain this fix.)

**Comment by crw** [comment]
Hi @thomasmeeus, I'm cleaning up old issues and I see this one did not receive a response. I am going to close this issue now, please let me know if you believe this action to be taken in error. Thanks!
